ddprof-lib/src/main/cpp/rustDemangler.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2021, 2023 Datadog, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "rustDemangler.h"
+
+#include <array>
+
+namespace RustDemangler {
+
+// With some exceptions we don't handle here, v0 Rust symbols can end in a
+// prefix followed by a 16-hexdigit hash, which must be removed
+const std::string hash_pre = "::h";
+const std::string hash_eg = "0123456789abcdef";
+
+const std::array<std::pair<std::string, std::string>, 9> patterns = {{
+    {"..", "::"},
+    {"$C$", ","},
+    {"$BP$", "*"},
+    {"$GT$", ">"},
+    {"$LT$", "<"},
+    {"$LP$", "("},
+    {"$RP$", ")"},
+    {"$RF$", "&"},
+    {"$SP$", "@"},
+}};
+
+inline bool is_hexdig(const char c) {
+  return (c >= 'a' && c <= 'f') || (c >= '0' && c <= '9');
+}
+
+// Simple conversion from hex digit to integer
+inline int hex_to_int(char dig) {
+  constexpr int k_a_hex_value = 0xa;
+  if (dig >= '0' && dig <= '9') {
+    return dig - '0';
+  }
+  if (dig >= 'a' && dig <= 'f') {
+    return dig - 'a' + k_a_hex_value;
+  }
+  if (dig >= 'A' && dig <= 'F') {
+    return dig - 'A' + k_a_hex_value;
+  }
+  return -1;
+}
+
+// Minimal check that a string can end, and does end, in a hashlike substring
+inline bool has_hash(const std::string &str) {
+  // If the size can't conform, then the string is invalid
+  if (str.size() <= hash_pre.size() + hash_eg.size()) {
+    return false;
+  }
+
+  // Check that the string contains the hash prefix in the right position
+  if (str.compare(str.size() - hash_eg.size() - hash_pre.size(),
+                  hash_pre.size(), hash_pre)) {
+    return false;
+  }
+
+  // Check that the string ends in lowercase hex digits
+  for (size_t i = str.size() - hash_eg.size(); i < str.size(); ++i) {
+    if (!is_hexdig(str[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool is_probably_rust_legacy(const std::string &str) {
+  // Is the string too short to have a hash part in thefirst place?
+  if (!has_hash(str)) {
+    return false;
+  }
+
+  // Throw out `$$` and `$????$`, but not in-between
+  const char *ptr = str.data();
+  const char *end = ptr + str.size() - hash_pre.size() - hash_eg.size();
+  for (; ptr <= end; ++ptr) {
+    if (*ptr == '$') {
+      if (ptr[1] == '$') {
+        return false;
+      }
+      return ptr[2] == '$' || ptr[3] == '$' || ptr[4] == '$';
```

Comment:
Does this mean that there will be at most 5 characters in the string?

Reply:
Sorry, I couldn't think of a way of implementing this section cleanly without resorting to patterns/lookups that would have bloated the code even more.

Overall, this function is trying to check whether a given string _might_ be rust-legacy-mangled.  For that to be the case, it needs to end in a "hash" (checked separately above), and if it has a `$`, then the first observation thereof has to be in a valid pattern (`$?$` or `$??$` or `$???$`, but not `$$` or `$` (alone) or `$????$` etc).  So rather than there just being 5 characters, this is actually a fast sieve for checking that if there is a `$` then the first instance thereof is within a valid pattern.

It could be the case that the string is something like `hello$GT$wo$$::h0123456789abcdef`, which would pass as _probably_ rust legacy (even though it isn't) and get partially "demangled."  This implementation uses the sieve as a good-enough heuristic, since it'd be reeeeaally weird otherwise.

Reply:
Just to be clear, we could absolutely check that all patterns are valid.  In fact, we'd probably just drop the check and then implement the replacement in a way that it returns error if the conversion hit an invalid parse.

The problem is it'd just be more expensive.